File: hw1/scala/task3.scala
import org.apache.spark.{HashPartitioner, SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, PrintWriter}
import scala.util.parsing.json.JSON

object task3 {
  def main(args: Array[String]): Unit = {

    val input_file = args(0)//"E:\\Data\\inf553\\hw1\\review.json"
    val output_file = args(1)//"task3_default_ans"
    val partition_type = args(2)//"customized" //default
    val n_partitions = args(3)//"20"
    val n = args(4)//"50"

    val conf = new SparkConf().setMaster("local[*]").setAppName("task1")
    val sc = new SparkContext(conf)

    var rdd = sc.textFile(input_file).map(item => parse(item)).map(item => (compact(item \ "business_id"), 1))
    //    var rdd = sc.textFile(input_file).map(JSON.parseFull)
    //      .map(item => item.get.asInstanceOf[Map[String, Any]]).map(item => (item("business_id"), 1))

    if (partition_type.trim() == "customized") {
      rdd = rdd.partitionBy(new HashPartitioner(n_partitions.toInt))
    }
    val n_par = rdd.getNumPartitions
    val n_items = rdd.glom().map(_.length).collect().toList
    val result = rdd.reduceByKey(_ + _).filter(item => item._2 > n.toInt).collect().toList
    val pw = new PrintWriter(new File(output_file))
    pw.write("{\"n_partitions\": " + n_par + ", ")
    pw.write("\"n_items\": [" + n_items.mkString(",") + "], ")
    pw.write("\"result\": [")
    result.foreach(item => {
      pw.write("[" + item._1 + ", " + item._2 + "]")
      if (item._1 != result.last._1) {
        pw.write(", ")
      }
    })
    pw.write("]}")
    pw.close()
  }
}

File: hw1/python/task3.py
import sys
import json
from pyspark import SparkContext

if __name__ == '__main__':

    # input_file, output_file, partition_type, n_partitions, n = "review.json", "default_ans", "customized", 20, 50
    input_file, output_file, partition_type, n_partitions, n = sys.argv[1:]
    n_partitions, n = int(n_partitions), int(n)
    sc = SparkContext("local[*]", 'task3')
    rdd = sc.textFile(input_file).map(json.loads).map(lambda x: [x['business_id'], 1])
    if partition_type.strip() == 'customized':
        rdd = rdd.partitionBy(n_partitions, hash)
    rdd1 = rdd.reduceByKey(lambda a, b: a + b).filter(lambda x: x[1] > n).map(list)
    result = dict()
    result['n_partitions'] = rdd.getNumPartitions()
    result['n_items'] = rdd.glom().map(len).collect()
    result['result'] = rdd1.collect()
    with open(output_file, 'w') as f:
        json.dump(result, f)